inspirehep/migrator/tasks.py

```diff
@@ -126,54 +104,75 @@ def migrate_record_from_legacy(recid):
     elif also_migrate == "broken":
         query = query.filter(LegacyRecordsMirror.valid.isnot(True))
     elif also_migrate != "all":
         raise ValueError('"also_migrate" should be either None, "all" or "broken"')
 
-    if wait_for_results:
-        # if the wait_for_results is true we enable returning results from the
-        # migrate_recids_from_mirror task so that we could use them to
-        # synchronize migrate task (which in that case waits for the
-        # migrate_recids_from_mirror tasks to complete before it finishes).
-        tasks = []
-        migrate_recids_from_mirror.ignore_result = False
-
-    chunked_recids = chunker(res.recid for res in query.yield_per(CHUNK_SIZE))
-    for i, chunk in enumerate(chunked_recids):
-        scheduled_records = i * CHUNK_SIZE + len(chunk)
-        print(f"Scheduled {scheduled_records} records for migration")
-        if wait_for_results:
-            tasks.append(migrate_recids_from_mirror.s(chunk))
-        else:
-            migrate_recids_from_mirror.delay(chunk)
+    recids_chunked = chunker(res.recid for res in query.yield_per(CHUNK_SIZE))
+    task = migrate_recids_from_mirror(
+        recids_chunked, disable_orcid_push=disable_orcid_push
+    )
+    print("All migration tasks has been scheduled.")
+    return task
+
+
+@shared_task(ignore_results=False, queue="migrator", acks_late=True)
+def migrate_recids_from_mirror(recids_chunks, step_no=0, disable_orcid_push=True):
+    """Task to migrate a chunked list of recids from the mirror.
+        recid_chunks (generator): Generator of record ids chunked for workers to pick.
+        step_no (int): Current step in `migration_steps`
+        disable_orcid_push (bool): flag indicating whether the orcid_push
+            should be disabled (if True) or executed at the end of migrations (if False).
 
-    if wait_for_results:
-        job = group(tasks)
-        result = job.apply_async()
-        result.join()
-        migrate_recids_from_mirror.ignore_result = True
-        print("All migration tasks have been completed.")
+    """
+    migration_steps = [
+        create_records_from_mirror_recids,
+        recalculate_citations,
+        index_records,
+    ]
+    if not disable_orcid_push:
+        migration_steps.append((run_orcid_push,))
+
+    if step_no >= len(migration_steps):
+        return
+
+    LOGGER.info(f"Running migration step {step_no+1}/{len(migration_steps)}")
+
+    all_recs = 0
+    for row in recids_chunks:
+        all_recs += len(row)
+    if all_recs == 0:
+        LOGGER.warning("There are no records to migrate. Terminating.")
+        return None
+    LOGGER.info(f"Processing {all_recs} records in this migration step")
+    step = migration_steps[step_no]
+    header = (step.s(r) for r in recids_chunks)
+    chord_task = chord(header)(
+        migrate_recids_from_mirror.s(
+            step_no=step_no + 1, disable_orcid_push=disable_orcid_push
+        )
+    )
```

Comment:
This triggers a task even if you are at the last step. What about
```
if step_no < len(migration_steps) -1:
    chord_task = chord(header)...
```
?

Reply:
I explained this IRL